### PR TITLE
Token Burn Precompile Refactoring

### PR DESCRIPTION
**Pull request by CordonaCodeCraft**
This PR proposes refactoring of the TokenBurnTransitionLogic in order to get the service compatible with the Precompile operation. 
The refactoring is only partial and is by no means a comprehensive solution.
In its current state it only introduces a new class (BurnLogic) and a transfer of the business and validation logic, previously implemented in the transition logic itself. 
E2E tests are not being verified as passing, unit tests are failing.
Those will be fixed as soon as the service is entirely refactored and utilizing new classes like SideEffectsTracker.

### CODE REVIEW

File: hedera-node/src/main/java/com/hedera/services/store/TypedTokenStore.java [hunk 1]
```diff
@@ -119,6 +119,7 @@ public TypedTokenStore(
 		this.sideEffectsTracker = sideEffectsTracker;
 		this.backingTokenRels = backingTokenRels;
 		this.addKnownTreasury = legacyStoreDelegate;
+
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/TypedTokenStore.java:122]: nit: Unnecessary change

File: hedera-node/src/main/java/com/hedera/services/txns/token/TokenBurnTransitionLogic.java [hunk 2]
```diff
@@ -77,26 +75,12 @@ public void doStateTransition() {
...
-			tokenStore.loadUniqueTokens(token, burnList);
-			token.burn(ownershipTracker, treasuryRel, burnList);
-		}
-
-		/* --- Persist the updated models --- */
-		tokenStore.commitToken(token);
-		tokenStore.commitTokenRelationships(List.of(treasuryRel));
-		tokenStore.commitTrackers(ownershipTracker);
-		accountStore.commitAccount(token.getTreasury());
+		final var burnLogic = new TokenBurnLogic();
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/txns/token/TokenBurnTransitionLogic.java:82]: This should be injected into the constructor as we have it injected in the `Token Mint PR`

File: hedera-node/src/main/java/com/hedera/services/txns/token/operation/TokenBurnLogic.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
+package com.hedera.services.txns.token.operation;
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/txns/token/operation/TokenBurnLogic.java:1]: nit: License

File: hedera-node/src/main/java/com/hedera/services/txns/token/operation/TokenBurnLogic.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import com.hedera.services.store.AccountStore;
+import com.hedera.services.store.TypedTokenStore;
+import com.hedera.services.store.models.Id;
+import com.hedera.services.store.models.OwnershipTracker;
+
+import java.util.List;
+
+public class TokenBurnLogic {
+
+	public void burn(
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/txns/token/operation/TokenBurnLogic.java:13]: `tokenStore` and `accountsStore` should be part of the properties in the `TokenBurnLogic` and be passed in the constructor

**Review by Daniel-K-Ivanov — CHANGES_REQUESTED** [review]

File: hedera-node/src/main/java/com/hedera/services/txns/token/BurnLogic.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+import com.hedera.services.state.enums.TokenType;
+import com.hedera.services.store.AccountStore;
+import com.hedera.services.store.TypedTokenStore;
+import com.hedera.services.store.models.Id;
+import com.hedera.services.store.models.OwnershipTracker;
+
+import java.util.List;
+
+public class BurnLogic {
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/txns/token/BurnLogic.java:31]: I think that we should make this injectable as MintLogic: https://github.com/hashgraph/hedera-services/pull/2516/files#diff-e50754ffc994d0767f46b936ffe56a9948363bdc519b0b3b7044f24884fd13e1R43

**Review by Daniel-K-Ivanov — CHANGES_REQUESTED** [review]

**Review by Daniel-K-Ivanov — APPROVED** [review]

**Review by stoqnkpL — APPROVED** [review]
LGTM
